apps/prairielearn/src/tests/render.test.ts

```diff
@@ -0,0 +1,454 @@
+import { dirname, join, resolve } from 'path';
+import { fileURLToPath } from 'url';
+
+import { config as chaiConfig } from 'chai';
+import { assert } from 'chai'; // Import assert
+// import chai from 'chai';
+import fs from 'fs-extra';
+import { HTMLRewriter } from 'html-rewriter-wasm';
+import { HtmlValidate } from 'html-validate';
+import { HTMLHint } from 'htmlhint';
+import * as log from 'why-is-node-running';
+
+import * as assetServer from '../lib/assets.js';
+import * as codeCaller from '../lib/code-caller/index.js';
+import { config } from '../lib/config.js';
+import type { Course, Question, Submission, Variant } from '../lib/db-types.js';
+import { features } from '../lib/features/index.js';
+import * as loadEstimator from '../lib/load.js';
+import { buildQuestionUrls } from '../lib/question-render.js';
+import { makeVariant } from '../lib/question-variant.js';
+import * as freeformServer from '../question-servers/freeform.js';
+import * as questionServers from '../question-servers/index.js';
+import * as helperServer from '../tests/helperServer.js';
+const htmlvalidate = new HtmlValidate();
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = dirname(__filename);
+
+const exampleCoursePath = resolve(__dirname, '..', '..', '..', '..', 'exampleCourse');
+const questionsPath = join(exampleCoursePath, 'questions');
+chaiConfig.showDiff = true;
+
+const buildSubmission = (answer: any): any => {
+  // hacky way to recover the expected input given the correct answer.
+  for (const key in answer) {
+    // if array, loop over
+    if (Array.isArray(answer[key])) {
+      answer[key] = answer[key].map((item) => {
+        if (typeof item === 'object' && 'key' in item) {
+          return item.key;
+        }
+        return item;
+      });
+      if (answer[key].some((v) => typeof v === 'object' && 'inner_html' in v)) {
+        answer[`${key}-input`] = JSON.stringify(answer[key]);
+        delete answer[key];
+      } else if (answer[key].some((v) => typeof v === 'object' && ('top' in v || 'x1' in v))) {
+        // pl-drawing
+        answer[key] = JSON.stringify(answer[key]);
+      } else {
+        let i = 0;
+        for (const v in answer[key]) {
+          // matching
+          answer[`${key}-dropdown-${i}`] = v;
+          i += 1;
+        }
+      }
+      // pl-drawing
+      // if (answer[key].length === 0) {
+      //   answer[key] = '[]';
+      // }
+    } else if (typeof answer[key] === 'number') {
+      // if number, convert to string
+      answer[key] = answer[key].toString();
+    } else if (typeof answer[key] === 'object') {
+      if ('_type' in answer[key] && answer[key]._type === 'ndarray') {
```

Comment:
Need to provide the submission by key for some elements, and as an array for others.